machine/dni/dni_3048up/u-boot/platform-add-recovery-mode-option.patch

```diff
@@ -0,0 +1,1693 @@
+@@ -2025,6 +3461,9 @@ void iproc_save_shmoo_values(void)
+         printf("Error allocating memory for saving Shmoo values!\n");
+         return;
+     }
++#if defined(CONFIG_DNI_3048UP_RECOVERY)
++    printf("%s: length=%d, address=0x%08X\n", __func__, length, buffer);
++#endif
+     ptr = buffer;
+ 
+     /* Fill signature */
+@@ -2774,7 +4213,9 @@ int ddr_init2(void)
+   {  /* this statement is to reduce compile error */
+   }
+ #else
++#if !defined(CONFIG_DNI_3048UP_RECOVERY)
+   while(!(reg32_read((volatile uint32_t *)DDR_DENALI_CTL_89) & 0x100));
++#endif
+ #if defined(IPROC_DDR_INIT2_DEBUG)
+ 	printf("ddr_init2: MemC initialization complete\n");
+ #endif
+diff --git a/arch/arm/include/asm/iproc/ddr_bist.h b/arch/arm/include/asm/iproc/ddr_bist.h
+index f5c1183..55617e9 100755
+--- a/arch/arm/include/asm/iproc/ddr_bist.h
++++ b/arch/arm/include/asm/iproc/ddr_bist.h
+@@ -17,8 +17,13 @@ extern uint32_t iproc_get_ddr3_clock_mhz(uint32_t unit);
+ #define SOC_DDR3_CLOCK_MHZ(unit)	iproc_get_ddr3_clock_mhz(unit)
+ 
+ #if (defined(CONFIG_HELIX4) || defined(CONFIG_KATANA2))
+-#define DDR_SHMOO_PARAM_MEM_PTR		(0x1b040000) /*size ~0x300 */
+-#define DDR_SHMOO_VREFW_MEM_PTR		(0x1b041000) /* size ~0x30000 */
++#  if defined(CONFIG_DNI_3048UP_RECOVERY)
++#    define DDR_SHMOO_PARAM_MEM_PTR     (0x1b05B000) /* use top of internal SRAM */
++#    define DDR_SHMOO_VREFW_MEM_PTR     (0x1b05C000)
++#  else
++#    define DDR_SHMOO_PARAM_MEM_PTR		(0x1b040000) /*size ~0x300 */
++#    define DDR_SHMOO_VREFW_MEM_PTR		(0x1b041000) /* size ~0x30000 */
++#  endif
+ #elif defined(CONFIG_HURRICANE2)
+ #define DDR_SHMOO_PARAM_MEM_PTR		(0x1b000000)
+ #define DDR_SHMOO_VREFW_MEM_PTR		(0x50000000)
+diff --git a/board/broadcom/dni_3048up/config.mk b/board/broadcom/dni_3048up/config.mk
+index 98845cc..295b662 100644
+--- a/board/broadcom/dni_3048up/config.mk
++++ b/board/broadcom/dni_3048up/config.mk
+@@ -8,6 +8,10 @@ ifdef CONFIG_IPROC_SPL
+ CONFIG_SYS_TEXT_BASE = 0x80000000
+ endif
+ 
++ifdef CONFIG_DNI_3048UP_RECOVERY
++CONFIG_SYS_TEXT_BASE = 0x1B008000
++endif
++
+ ifdef CONFIG_NO_CODE_RELOC
+ ifdef CONFIG_SYS_BIG_ENDIAN
+ LDSCRIPT := $(SRCTREE)/board/$(BOARDDIR)/u-boot-nr-be.lds
+diff --git a/board/broadcom/dni_3048up/dni_3048up.c b/board/broadcom/dni_3048up/dni_3048up.c
+index 33c2e70..b9042ea 100644
+--- a/board/broadcom/dni_3048up/dni_3048up.c
++++ b/board/broadcom/dni_3048up/dni_3048up.c
+@@ -346,7 +346,9 @@ int board_late_init (void)
+     buffer = (buffer&0xfc);  /* turn off bits 1:0 */
+     i2c_write(0x28, 0x0a, 1, &buffer, 1);
+ 	
++#ifdef CONFIG_CMD_USB
+ 	linux_usbh_init();
++#endif
+ 
+ #if 0
+ #ifdef CONFIG_POPULATE_SERIAL_NUMBER
+diff --git a/board/broadcom/dni_3048up/usb.c b/board/broadcom/dni_3048up/usb.c
+index c1f6d26..7cda2c8 100644
+--- a/board/broadcom/dni_3048up/usb.c
++++ b/board/broadcom/dni_3048up/usb.c
+@@ -18,6 +18,8 @@
+  * MA 02111-1307 USA
+  */
+ 
++#ifdef CONFIG_CMD_USB
++
```

Comment:
This is not quite right.  This `#ifdef` needs to come after `#include <common.h>`.